File: Current version/c++/PluginAdaptor/RESTHelper.cc
#include "RESTHelper.h"
#include "Headers.h"


namespace We3Graph
{
	json::value RESTHelper::SendTypicalRequest(string url,
		unordered_map<string, string> & urlParameters, string method,
		json::value & jsonBody)
	{
		uri_builder uri(Statics::StringToWide(url));

		if (method == "Get" || method == "Delete")
		{
			for (auto item : urlParameters)
				uri.append_query(Statics::StringToWide(item.first),
				Statics::StringToWide(item.second));

		}

		auto httpClient = http_client(uri.to_string());
		auto httpRequest = http_request();


		if (method == "Post" || method == "Put" || method == "Delete")
		{
			if (method == "Post")
				httpRequest.set_method(methods::POST);
			else if (method == "Put")
				httpRequest.set_method(methods::PUT);
			else if (method == "Delete")
				httpRequest.set_method(methods::DEL);

			auto serializedBody = jsonBody.serialize();
			httpRequest.set_body(serializedBody, L"application/json");
		}
		else
			httpRequest.set_method(methods::GET);

		auto responseTask = httpClient.request(httpRequest);
		responseTask.wait();
		auto response = responseTask.get();

		if (response.status_code() == status_codes::OK)
		{
			auto responseJSONTask = response.extract_json();
			responseJSONTask.wait();
			return responseJSONTask.get();
		}

		else if (response.status_code() == status_codes::Unauthorized)
			throw new AuthenicationException();
		else if (response.status_code() == status_codes::Forbidden)
			throw new AuthorizationException();
		else
			throw new WebServiceException();
	}
}

File: Thesis version/c#/PluginAdaptor/RESTHelper.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using System.Web.Script.Serialization;

namespace We3Graph
{
    /// <summary>
    /// A static class to help send REST requests
    /// </summary>
    class RESTHelper
    {
        public static object SendTypicalRequest(string url,
            Dictionary<string, string> urlParameters, string method,
            object bodyParameters = null)
        {
            JavaScriptSerializer serializer = new JavaScriptSerializer();

            using (var webClient = new WebClient())
            {
                string responseString = "";
                string jsonBody = "";

                if (method == "Get" || method == "Delete")
                {
                    StringBuilder sb = new StringBuilder(url);

                    bool isFirst = true;
                    foreach (var item in urlParameters)
                    {
                        if (isFirst)
                            sb.Append("?");
                        else
                            sb.Append("&");
                        isFirst = false;

                        sb.Append(Uri.EscapeDataString(item.Key));
                        sb.Append("=");
                        sb.Append(Uri.EscapeDataString(item.Value));
                    }

                    url = sb.ToString();
                }

                if (method == "Post" || method == "Put" || method == "Delete")
                {
                    webClient.Headers["Content-Type"] = "application/json";
                    if (bodyParameters == null)
                        bodyParameters = new object { };
                    jsonBody = serializer.Serialize(bodyParameters);
                }

                try
                {
                    if (method == "Post" || method == "Put" || method == "Delete")
                        responseString = webClient.UploadString(url, method, jsonBody);
                    else if (method == "Get")
                        responseString = webClient.DownloadString(url);
                }
                catch (WebException e)
                {
                    HttpStatusCode statusCode = ((HttpWebResponse)e.Response).StatusCode;
                    string response = "";
                    using (var reader = new StreamReader(e.Response.GetResponseStream()))
                    { response = reader.ReadToEnd(); }

                    if (statusCode == HttpStatusCode.Unauthorized)
                        throw new AuthenicationException();
                    if (statusCode == HttpStatusCode.Forbidden)
                        throw new AuthorizationException();
                    else
                        throw new WebServiceException();
                }

                return serializer.DeserializeObject(responseString);
            }
        }

    }
}
